### ISSUE TITLE
cmd/govim: investigate prop_add flake

### ISSUE THREAD

**Issue by myitcv** [govim/govim#655]
From a recent CI run. The test itself doesn't appear to be significant, except for the fact that line removal is probably involved:

```
--- FAIL: TestScripts (43.60s)
    main_test.go:63: using gopls at "/tmp/gobin-gopls-installdir074416177/gopls"
    --- FAIL: TestScripts/scenario_default (0.00s)
        --- FAIL: TestScripts/scenario_default/goimports (1.43s)
            testscript.go:337: 
                # Test that govim#config#Set("FormatOnSave", "goimports") and GOVIMGoImports work
                # Notice below: file.go and file.go.orig are identical. This is so that we
                # don't need to copy file.go.orig to file.go before opening file.go Because
                # doing so triggers govim to consider the file change from a file watcher
                # perspective which throws the observed diagnostics etc.
                # :GOVIMGoImports whole file (0.594s)
                # Format on save (0.313s)
                # Disabled pending resolution to https://github.com/golang/go/issues/34103
                # errlogmatch -count=0 'LogMessage callback: &protocol\.LogMessageParams\{Type:(1|2), Message:".*'
                # Format on save (bad syntax) (0.045s)
                > cp file.go.bad file.go
                > vim ex 'e! file.go'
                [stderr]
                unexpected command error: failed to ex(e! file.go) in Vim: Caught 'got error whilst handling autocommand:1: driver error: ChannelCall("s:batchCall") failed: failed to call s:batchCall([[[call s:mustNothing prop_remove {0 1 1}] [call s:mustNothing prop_add 4 5 {GOVIMErr 4 21 1}]]]) in Vim: Caught ''Vim(let):E966: Invalid line number: 4'' in function <SNR>22_callbackAutoCommand[18]..<SNR>22_ch_evalexpr[2]..<SNR>22_define[51]..<SNR>22_batchCall, line 21' in function <SNR>22_callbackAutoCommand[18]..<SNR>22_ch_evalexpr, line 4
                [exit status 255]
                FAIL: testdata/scenario_default/goimports.txt:29: unexpected command failure
                
FAIL
FAIL	github.com/govim/govim/cmd/govim	149.698s
```

i.e. we try to add props to a line that no longer exists?

**Comment by myitcv** [comment]
Logs available from a recent build fail via:

```
cd $(mktemp -d) && curl -s https://gist.githubusercontent.com/myitcv/28b3f00d5f1a066fee2be7de61c41084/raw/ece0c49cea78e06b69a0d283d73a6d2d4165b15c/logs.base64 | base64 -d | tar -zx
```

**Comment by leitzler** [comment]
Ok, so what's happening is that we don't receive any diagnostics for one of the edits (`DidChange` version 7 in the logs above).

We then re-open the file `e! file.go`, but this time with new content. Vim sends a `BufUnloaded` and 
`BufNewFile/BufRead` for that buffer (still buf number 1).

Since the buffer already exist in govim (it isn't and shouldn't be dropped at `BufUnloaded`) we just reload the file and try to place highlights according to the last known diagnostics.

Text properties are automatically removed when a buffer is unloaded, so we must be able to place them again when a buffer is loaded (note that unload/load in vim happens without the file being closed, i.e. will happen when jumping between open buffers).

We should avoid placing text properties when the buffer content has changed since it was unloaded.

**Comment by myitcv** [comment]
> We should avoid placing text properties when the buffer content has changed since it was unloaded.

Isn't the safest approach here to only place text properties when the diagnostics version matches the buffer version?

**Comment by leitzler** [comment]
Yes, but since it's `govim` that increases version I think that we might have to compare buffer contents to be able to avoid increasing version when the contents hasn't changed.

Currently we (incorrectly?) blindly increase version whenever we get a `BufReadPost`, i.e. if you have `a.go` and `b.go` open in separate buffers and switch between them back and forth we will increase version and send `didChange` to gopls each time regardless of contents. I think we have to avoid that since gopls shouldn't(?) send diagnostics if the diagnostics hasn't changed.

**Comment by myitcv** [comment]
Copying in our exchange on Slack here for context:

> > Regarding when diagnostics should be sent by gopls and not, what's the latest take if the client send a DidChange that doesn't alter any diagnostics. Should gopls still send the same diagnostics but with increased version to "acknowledge" the change?
>>
>> I think I've read that diagnostics shouldn't be send if they didn't change.
>
> If the version number for a file changes diagnostics have to be resent AFAIK
> 
> Otherwise the diagnostics for a file are not resent if they haven't changed

`govim` is the source of truth for version numbers. So if we bump the version number, regardless of whether the content changes, `gopls` has to follow suit.

My understanding is that `gopls` always sends diagnostics for a file when its version number increases, and only sends diagnostics if they change otherwise (i.e. no change in version number). Anything else is a bug.

Therefore it's an optimisation for `govim` not to send `DidChange` version increments if the contents haven't changed, but it's not essential. But it's an optimisation well worth doing, per your PR in #665.

**Comment by leitzler** [comment]
I haven't seen this since #665 was merged, and it cannot happen after #741 is merge since this error from vim will be suppressed. Closing.
